Enable "portmap" plugin by default

klausenbusk opened the pull request:
~~Not tested (yet) and https://github.com/coreos/flannel-cni/pull/3 should be merged first.~~

~~Config taken from here: https://github.com/kubernetes-incubator/kube-aws/issues/704#issuecomment-322051829 , ~~he say it isn't working, but that could be because of the [extension](https://github.com/kubernetes/kubernetes/issues/23920#issuecomment-308254308).~~ **Update:** It is [working](https://github.com/kubernetes-incubator/kube-aws/issues/704#issuecomment-323609993)~~

chadswen commented on install-cni.sh:
```diff
@@ -7,16 +7,16 @@ if [ -w "/host/opt/cni/bin/" ]; then
 fi
 
-FILENAME=${CNI_CONF_NAME:-10-flannel.conf}
+FILENAME=${CNI_CONF_NAME:-10-flannel.conflist}
```
Would a residual `10-flannel.conf` from a previous installation conflict with `10-flannel.conflist` after upgrading? If supporting upgrades is in scope right now we probably want to remove or rename the old default file.

blu3gui7ar replied:
Force the cp to ensure overwrite would be better.
cp -f /opt/cni/bin/* /host/opt/cni/bin/;

And also update Dockerfile
ADD flannel.conf.default /flannel.conf.default

klausenbusk commented on install-cni.sh:
```diff
@@ -7,16 +7,16 @@ if [ -w "/host/opt/cni/bin/" ]; then
 fi
 
-FILENAME=${CNI_CONF_NAME:-10-flannel.conf}
+FILENAME=${CNI_CONF_NAME:-10-flannel.conflist}
```
You correct :)
I will add a `rm` to the script.

klausenbusk commented:
@chadswen @blu3gui7ar I have tested this now on my bootkube cluster, and it is working (yeh!).

dghubble commented:
We need to be sure this does not interact badly with clusters that have the old config files in place. Prior clusters and migrations need to be tested.

klausenbusk commented:
> We need to be sure this does not interact badly with clusters that have the old config files in place. Prior clusters and migrations need to be tested.

Yeh, the file extensions was changed from `.conf` to `.conflist`, and I'm not sure a `.conf` will work with a `.conflist` extension.

`.conflist` can have multiple plugins, `.conf` can't and I don't think `.conflist` support specifying the plugin as `type`.

I'm not sure how this should be managed.

klausenbusk commented:
Looking at the CNI code, the two extension are handled totally different:
https://github.com/containernetworking/cni/blob/80c688516db9873ba8b9d630014dadde5608dc42/libcni/conf.go#L43

https://github.com/containernetworking/cni/blob/80c688516db9873ba8b9d630014dadde5608dc42/libcni/conf.go#L62

Ideally we should support both type of config, but I'm not sure how.

klausenbusk commented:
Any update on this?

For people who isn't using `CNI_NETWORK_CONFIG`, it should just work. People using `CNI_NETWORK_CONFIG` need to update their config ( `bootkube`/`tectonic` users??).
People shouldn't upgrade a component without checking the changelog.

A big warning in the changelog, shouldn't that be enough?

dghubble commented:
The `flannel-cni` container just populates `/etc/cni/net.d/$CNI_CONF_NAME` with the contents of `$CNI_NETWORK_CONFIG`. With that in mind, it is fine to change the default conf name and config contents, as long as users can still set the old name and old config to retain current behavior.

### Compatibility

(Desired) Preserve the current behavior by explicitly setting the `CNI_CONFIG_NAME` (`10-flannel.conf` or whatever custom name you're using) as an env variable to `install-cni`.

```
- name: install-cni
  image: quay.io/coreos/flannel-cni:0.1
  command: ["/install-cni.sh"]
  env:
   - name: CNI_CONF_NAME
     value: "10-flannel.conf"
   - name: CNI_NETWORK_CONFIG
     valueFrom:
     configMapKeyRef:
        name: kube-flannel-cfg
        key: cni-conf.json
```

The new `install-cni` will remove the config temporarily, but should recreate it with the same contents.

### Fixes Needed

Let's not assume the file to be removed is `/etc/cni/net.d/10-flannel.conf`. Perhaps, respect a `CNI_OLD_NAME` environment variable (default to `10-flannel.conf`) and remove that file.

Benefits:

* Avoids breaking anyone who wants to keep using their `10-flannel.conf` setup (set the name explicitly via environment variable, it'll remove and recreate)
* Provides a generic way to cleanup the previous CNI config (say someday we change the default again from `10-flannel.conflist` to something else, then we default CNI_OLD_NAME to 10-flannel.conflist)
* Supports users who are already customizing the name to something other than `10-flannel.conf`
* Supports users who want to switch to `10-flannel.conflist` out of the box
* Allows users to do their own migrations between discrete, cluster-wide flannel CNI configs

issue: #4 
upgrade example: https://github.com/kubernetes-incubator/bootkube/pull/697

Note: A warning in a changelog and requiring users to change their configs in order to upgrade to the next flannel-cni release isn't the right way to think about this. Tectonic promises seamless updates so they would need to create a way for their operators to migrate these configs which may be tricky to coordinate and slow an update in the event of a vulnerability later. Its important to avoid breakage because one cannot foresee the cascading effects. Small changes are not small impact.

klausenbusk committed:
Switch to .conflist config as default

.conflist config is needed for more advanced configuration (ex:
plugin chaining, needed for portmap).
This commit also add a new env variable: CNI_OLD_NAME , which
specific the name of the old config, which is removed on install.
Default to 10-flannel.conf

klausenbusk committed:
Enable "portmap" plugin by default

Make stuff like hostPort work with CNI.

Fix #4

klausenbusk commented:
@dghubble how does this look?

dghubble reviewed (approved):
Thanks
